### PR TITLE
Handle Close requests for clients and send disconnect signal to broker

### PR DESCRIPTION
**Pull request by vipeller**

### CODE REVIEW

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs [hunk 1]
```diff
@@ -27,7 +28,7 @@ public async Task DirectMethodCallForwarded()
                 .Setup(h => h.CallDirectMethodAsync(It.Is<DirectMethodRequest>(m => m == request), It.Is<IIdentity>(i => i == identity)))
                 .Returns(Task.FromResult(new DirectMethodResponse("123", new byte[0], 200)));
 
-            var sut = new DeviceProxy(identity, twinHandler, m2mHandler, c2dHandler, directMethodHandler);
+            var sut = new DeviceProxy(identity, connectionHanlder, twinHandler, m2mHandler, c2dHandler, directMethodHandler);
```
Comment (and-rewsmith) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs:31]: typo

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/IConnectionRegistry.cs [hunk 2]
```diff
@@ -11,5 +11,7 @@ public interface IConnectionRegistry
     {
         Task<Option<IDeviceListener>> GetDeviceListenerAsync(IIdentity identity);
         Task<Option<IDeviceProxy>> GetDeviceProxyAsync(IIdentity identity);
+
```
Comment (and-rewsmith) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/IConnectionRegistry.cs:14]: nit: should we remove this whitespace?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs [hunk 3]
```diff
@@ -94,6 +99,20 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
             return false;
         }
 
+        public async Task CloseConnectionAsync(IIdentity identity)
+        {
+            var identityFound = false;
+            using (await this.guard.LockAsync())
```
Comment (and-rewsmith) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:105]: We need to lock this because something might get added to knownConnections?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs [hunk 4]
```diff
@@ -94,6 +99,20 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
...
+            {
+                identityFound = this.knownConnections.TryGetValue(identity, out var _);
+            }
+
+            if (identityFound)
+            {
+                await this.connector.SendAsync(TopicDisconnect, Encoding.UTF8.GetBytes($"\"{identity.Id}\""));
+            }
```
Comment (and-rewsmith) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:113]: If the identity isn't found should we log error?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            DeviceProxy GetProxy(IIdentity identity)
+            {
+                return new DeviceProxy(
+                                identity,
+                                sut,
+                                Mock.Of<ITwinHandler>(),
+                                Mock.Of<IModuleToModuleMessageHandler>(),
+                                Mock.Of<ICloud2DeviceMessageHandler>(),
+                                Mock.Of<IDirectMethodHandler>());
+            }
```
Comment (and-rewsmith) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionHandlerTests.cs:66]: Is defining the function this way (i.e. inside CloseSendsDisconnectSignal) better than having a separate function?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs [hunk 6]
```diff
@@ -49,16 +51,17 @@ public async Task DesiredUpdateCallForwarded()
                 .Setup(h => h.SendDesiredPropertiesUpdate(It.Is<IMessage>(m => m == desired), It.Is<IIdentity>(i => i == identity)))
                 .Returns(Task.CompletedTask);
 
-            var sut = new DeviceProxy(identity, twinHandler, m2mHandler, c2dHandler, directMethodHandler);
+            var sut = new DeviceProxy(identity, connectionHanlder, twinHandler, m2mHandler, c2dHandler, directMethodHandler);
 
             await sut.OnDesiredPropertyUpdates(desired);
 
-            Mock.Get(directMethodHandler).VerifyAll();
+            Mock.Get(twinHandler).VerifyAll();
```
Comment (and-rewsmith) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs:58]: Why is this changed to twinHandler?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs [hunk 7]
```diff
@@ -71,16 +74,17 @@ public async Task Cloud2DeviceMessagesForwarded()
                 .Setup(h => h.SendC2DMessageAsync(It.Is<IMessage>(m => m == message), It.Is<IIdentity>(i => i == identity)))
                 .Returns(Task.CompletedTask);
 
-            var sut = new DeviceProxy(identity, twinHandler, m2mHandler, c2dHandler, directMethodHandler);
+            var sut = new DeviceProxy(identity, connectionHanlder, twinHandler, m2mHandler, c2dHandler, directMethodHandler);
 
             await sut.SendC2DMessageAsync(message);
 
-            Mock.Get(directMethodHandler).VerifyAll();
```
Comment (and-rewsmith) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs:78]: Why this change?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/DeviceProxy.cs [hunk 8]
```diff
@@ -41,15 +44,15 @@ public class DeviceProxy : IDeviceProxy
 
         public IIdentity Identity { get; }
 
-        public Task CloseAsync(Exception ex)
+        public async Task CloseAsync(Exception ex)
```
Comment (and-rewsmith) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/DeviceProxy.cs:47]: Do we do anything with this exception?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs (hunk 3 above)
Comment (vadim-kovalyov) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:105]: What race conditions we are trying to prevent with the lock? knownConnections is a ConcurrentDictionary already

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs [hunk 9]
```diff
@@ -94,6 +99,20 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
...
+        {
+            var identityFound = false;
+            using (await this.guard.LockAsync())
+            {
+                identityFound = this.knownConnections.TryGetValue(identity, out var _);
+            }
+
+            if (identityFound)
+            {
+                await this.connector.SendAsync(TopicDisconnect, Encoding.UTF8.GetBytes($"\"{identity.Id}\""));
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:112]: do we need to remove from knownConnections?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/DeviceProxy.cs (hunk 8 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/DeviceProxy.cs:47]: added to the log

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/IConnectionRegistry.cs (hunk 2 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/IConnectionRegistry.cs:14]: I added because those two above belong to one group of operations and the one below to a different one

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs (hunk 3 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:105]: The story of the lock is that when the code updates the 'knowConnection', that needs a lock anyway (because it does multistep things. so a parallel call cannot mess it up) - so it would not be necessary using a concurrent dictionary (although I ended up using one, because it has a member normal dictionaries don't) And because of the lock on writes, I felt it is better to keep it symmetric, and lock the reads, too.   But yes, with the current ConcurrentDictionary the read operation would not be dangerous doing without a lock.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs (hunk 4 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:113]: I am not sure if that is an error. Can a device be gone by the time edgeHub realizes that something is wrong a calls Close()? I can add an 'Info'.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionHandlerTests.cs (hunk 5 above)
Comment (vipeller) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionHandlerTests.cs:66]: it can capture local variables (sut in this case), also in tests these helper functions many times are used only from a single method, so it is easier to see what serves what.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs (hunk 6 above)
Comment (vipeller) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs:58]: it was a copy-paste error from earlier and it did not call the "verifyall" on the set up mock object. Now it checks the right mock. Earlier the test passed only because nothing was setup on 'directMethodHandler', so verifyAll meant 'verify nothing' - which always passed.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs (hunk 9 above)
Comment (vipeller) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs:112]: No, I figured that after this point it is the turn of the broker to do something, and when it make the client disconnect, the notification will come in a second round, and the identity will be removed that time

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs (hunk 7 above)
Comment (vipeller) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/DeviceProxyTest.cs:78]: it was a copy-paste error from earlier and it did not call the "verifyall" on the set up mock object. Now it checks the right mock. Earlier the test passed only because nothing was setup on 'directMethodHandler', so verifyAll meant 'verify nothing' - which always passed.

**Review by vadim-kovalyov — APPROVED** [review]

**Review by varunpuranik — APPROVED** [review]

**Review by and-rewsmith — APPROVED** [review]
